GitUpKit/Core/GCRepository+Bare.m

```diff
@@ -362,27 +366,67 @@ - (GCCommit*)createCommitFromTree:(git_tree*)tree
                             count:(NSUInteger)count
                            author:(const git_signature*)author
                           message:(NSString*)message
+                       shouldSign:(BOOL)shouldSign
                             error:(NSError**)error {
   GCCommit* commit = nil;
   git_signature* signature = NULL;
 
   git_oid oid;
+  
+  GCConfigOption* shouldSignOption = [self readConfigOptionForVariable:@"commit.gpgsign" error:nil];
   CALL_LIBGIT2_FUNCTION_GOTO(cleanup, git_signature_default, &signature, self.private);
-  CALL_LIBGIT2_FUNCTION_GOTO(cleanup, git_commit_create, &oid, self.private, NULL, author ? author : signature, signature, NULL, GCCleanedUpCommitMessage(message).bytes, tree, count, parents);
+  
+  git_buf commitBuffer = GIT_BUF_INIT_CONST("", 0);
+  const char *gpgSignature = NULL;
+  if ([shouldSignOption.value isEqualToString:@"true"]) {
+    GCConfigOption* signingKeyOption = [self readConfigOptionForVariable:@"user.signingkey" error:nil];
+    
+    CALL_LIBGIT2_FUNCTION_GOTO(cleanupBuffer, git_commit_create_buffer, &commitBuffer, self.private, author ? author : signature, signature, NULL, GCCleanedUpCommitMessage(message).bytes, tree, count, parents);
+    gpgSignature = [self gpgSig:commitBuffer.ptr keyId:signingKeyOption.value];
+  }
+  
+  if (gpgSignature != NULL) {
+    CALL_LIBGIT2_FUNCTION_GOTO(cleanupBuffer, git_commit_create_with_signature, &oid, self.private, commitBuffer.ptr, gpgSignature, NULL);
+  } else {
+    CALL_LIBGIT2_FUNCTION_GOTO(cleanupBuffer, git_commit_create, &oid, self.private, NULL, author ? author : signature, signature, NULL, GCCleanedUpCommitMessage(message).bytes, tree, count, parents);
+  }
+  
   git_commit* newCommit = NULL;
   CALL_LIBGIT2_FUNCTION_GOTO(cleanup, git_commit_lookup, &newCommit, self.private, &oid);
   commit = [[GCCommit alloc] initWithRepository:self commit:newCommit];
 
+cleanupBuffer:
+  git_buf_free(&commitBuffer);
+  
 cleanup:
   git_signature_free(signature);
   return commit;
 }
 
+-(const char*)gpgSig:(const char*)body keyId:(NSString*)keyId {
+  GPGKey *key = nil;
+  if (keyId.length > 0) {
+    key = [GPGKey secretKeyForId:keyId];
+  }
+  if (key == nil) {
+    key = [[GPGKey allSecretKeys] firstObject];
+  }
+  if (key == nil) {
+    return NULL;
+  }
+
+  NSString* plainToSign = [[NSString alloc] initWithCString:body encoding:NSUTF8StringEncoding];
+  NSString* signature = [key signSignature:plainToSign];
```

Comment:
@lucasderraugh I would like to trigger a separate discussion regarding UI blocking calls.
I'm focusing on signing commits functionality in current PR. However, even with the partial support I already observe significant UI regression – the main queue gets blocked for a 0.2-0.3 second when password for the gpg key already stored the keychain. It even can be an infinite period of time when GPG Keychain presents screen for GPG Key Password input.
Do you have any UI/UX related ideas on how we can play around this issue?

Reply:
@lucasderraugh Any ideas?

Reply:
Sorry, didn't realize there was an outstanding question. Let me think about what we could do here and I'll get back to you soon.

Reply:
@lucasderraugh any updates? 🙂